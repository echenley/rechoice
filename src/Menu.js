// @flow

import * as React from 'react'
import cx from 'classnames'
import DefaultMenu from './DefaultMenu'
import type {
  ClassMap,
  Item as ItemType,
  ItemRenderer,
  MenuRenderer,
  LoadingState,
} from './types'

type Props = {
  classes: ClassMap,
  controlsId: string,
  focusedItemIndex: ?number,
  getItemId: (i: number) => string,
  inputValue: string,
  isItemSelected: (item: ItemType) => boolean,
  isOpen: boolean,
  itemKey: string,
  items: Array<ItemType>,
  loadingIndicator?: React.Node,
  onHoverItem: (i: number) => mixed,
  onSelect: (item: ItemType) => mixed,
  renderItem: ItemRenderer,
  renderMenu?: MenuRenderer,
  focusedItemRef: (ref: ?HTMLElement) => mixed,
  menuRef: (ref: ?HTMLElement) => mixed,
  state?: LoadingState,
}

class Menu extends React.Component<Props> {
  renderItems = () => {
    const {
      classes,
      focusedItemIndex,
      getItemId,
      inputValue,
      isItemSelected,
      itemKey,
      items,
      onHoverItem,
      onSelect,
      renderItem,
      focusedItemRef,
    } = this.props

    return items.map((item, i) => {
      const isSelected = isItemSelected(item)
      const isFocused =
        typeof focusedItemIndex === 'number' && i === focusedItemIndex

      const renderedItem = renderItem({
        inputValue,
        isFocused,
        isSelected,
        item,
      })

      const itemClasses = cx(classes.item, renderedItem.props.className, {
        [classes.focused]: isFocused,
        [classes.selected]: isSelected,
      })

      return React.cloneElement(renderedItem, {
        'aria-selected': isFocused,
        className: itemClasses,
        id: getItemId(i),
        key: item[itemKey],
        onClick: e => {
          if (renderedItem.props.onClick) {
            renderedItem.props.onClick(e)
          }

          onSelect(item)
        },
        onMouseEnter: () => onHoverItem(i),
        ref: isFocused ? focusedItemRef : null,
        role: 'option',
      })
    })
  }

  render() {
    const {
      classes,
      controlsId,
      inputValue,
      isOpen,
      loadingIndicator,
      renderMenu = DefaultMenu,
      menuRef,
      state = 'success',
    } = this.props

    const menu = renderMenu({
      children: this.renderItems(),
      className: classes.menu,
      listProps: {
        id: controlsId,
        role: 'listbox',
      },
      inputValue,
      isOpen,
      loadingIndicator,
      menuRef,
      state,
    })

    if (!menu) {
      return menu
    }

    if (menu.type === 'string') {
      return React.cloneElement(React.Children.only(menu), {
        ref: menuRef,
      })
    }

    return menu
  }
}

export default Menu
