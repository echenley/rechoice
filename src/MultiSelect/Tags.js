// @flow

import * as React from 'react'
import type { Item, TagRenderer } from '../types'

type Props = {
  className: string,
  itemKey: string,
  removeTag: (item: Item) => mixed,
  renderTag: TagRenderer,
  value: Array<Item>,
}

class Tags extends React.Component<Props> {
  renderTag = (item: Item) => {
    const { itemKey, removeTag, renderTag } = this.props

    const tag = renderTag({
      item,
      removeTag: () => removeTag(item),
    })

    return React.cloneElement(tag, {
      key: item[itemKey],
    })
  }

  render() {
    const { className, value } = this.props
    return <span className={className}>{value.map(this.renderTag)}</span>
  }
}

export default Tags
