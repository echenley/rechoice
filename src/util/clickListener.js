// @flow

export function addClickListener(cb: EventHandler) {
  document.addEventListener('mousedown', cb)
  document.addEventListener('touchstart', cb)
}

export function removeClickListener(cb: EventHandler) {
  document.removeEventListener('mousedown', cb)
  document.removeEventListener('touchstart', cb)
}
