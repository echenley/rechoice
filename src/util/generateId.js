// @flow

let i = 0

export default function generateId(prefix: string) {
  return `${prefix}-${i++}`
}
