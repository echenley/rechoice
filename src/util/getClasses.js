// @flow

import type { ClassMap } from '../types'

export default (classes: $Shape<ClassMap>) => ({
  disabled: 'disabled',
  focused: 'focused',
  input: 'rechoice-input',
  inputMirror: 'rechoice-input-mirror',
  inputWrapper: 'rechoice-input-wrapper',
  item: 'rechoice-item',
  menu: 'rechoice-menu',
  root: 'rechoice',
  selected: 'selected',
  tags: 'rechoice-tags',
  value: 'rechoice-value',
  ...classes,
})
