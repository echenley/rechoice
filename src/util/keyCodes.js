// @flow

export default {
  BACKSPACE: 8,
  DELETE: 46,
  ENTER: 13,
  ESCAPE: 27,
  UP: 38,
  DOWN: 40,
  SPACEBAR: 32,
}
