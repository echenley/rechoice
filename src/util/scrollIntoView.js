// @flow

type NodeFinder = (node: HTMLElement) => boolean

export function findParent(
  finder: NodeFinder,
  node: ?HTMLElement,
  rootNode: HTMLElement
) {
  if (node && node !== rootNode.parentNode) {
    if (finder(node)) {
      return node
    }

    return findParent(finder, (node.parentNode: ?$Subtype<Node>), rootNode)
  }

  return null
}

const isScrollable = node => node.scrollHeight > node.clientHeight
const getClosestScrollParent = findParent.bind(null, isScrollable)

export default function scrollIntoView(
  node: HTMLElement,
  rootNode: HTMLElement
) {
  const scrollParent = getClosestScrollParent(node, rootNode)

  if (scrollParent === null) {
    return
  }

  const scrollParentStyles = getComputedStyle(scrollParent)
  const scrollParentRect = scrollParent.getBoundingClientRect()
  const scrollParentBorderTopWidth = parseInt(
    scrollParentStyles.borderTopWidth,
    10
  )
  const scrollParentBorderBottomWidth = parseInt(
    scrollParentStyles.borderBottomWidth,
    10
  )
  const scrollParentTop = scrollParentRect.top + scrollParentBorderTopWidth
  const nodeRect = node.getBoundingClientRect()
  const nodeOffsetTop = nodeRect.top + scrollParent.scrollTop
  const nodeTop = nodeOffsetTop - scrollParentTop

  if (nodeTop < scrollParent.scrollTop) {
    // item is above the scrollable area
    scrollParent.scrollTop = nodeTop
    return
  }

  const nodeBottom = nodeTop + nodeRect.height
  const totalScrollBorderHeight =
    scrollParentBorderTopWidth + scrollParentBorderBottomWidth

  if (
    nodeBottom + totalScrollBorderHeight >
    scrollParent.scrollTop + scrollParentRect.height
  ) {
    // item is below the scrollable area
    scrollParent.scrollTop =
      nodeTop +
      nodeRect.height -
      scrollParentRect.height +
      scrollParentBorderTopWidth +
      scrollParentBorderBottomWidth
  }
  // item is within the scrollable area (do nothing)
}
