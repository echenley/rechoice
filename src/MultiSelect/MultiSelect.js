// @flow

import * as React from 'react'
import cx from 'classnames'
import Menu from '../Menu'
import contains from '../util/contains'
import { addClickListener, removeClickListener } from '../util/clickListener'
import keyCodes from '../util/keyCodes'
import generateId from '../util/generateId'
import getClasses from '../util/getClasses'
import scrollIntoView from '../util/scrollIntoView'
import {
  focusInput,
  focusPreviousItem,
  focusNextItem,
  updateInputValue,
} from '../util/state'
import Tags from './Tags'
import type {
  ClassMap,
  FilterItemsMulti,
  Item,
  ItemRenderer,
  MenuRenderer,
  LoadingState,
  TagRenderer,
} from '../types'

type Props = {
  classes?: $Shape<ClassMap>,
  disabled: boolean,
  filterItems: FilterItemsMulti,
  inputRef?: (ref: ?HTMLInputElement) => mixed,
  isOpen?: boolean,
  itemKey: string,
  items: Array<Item>,
  loadingIndicator?: React.Node,
  loadingState?: LoadingState,
  minimumInput: number,
  onBlur?: () => mixed,
  onChange: (value: Array<Item>) => mixed,
  onFocus?: () => mixed,
  onInputChange?: (inputValue: string) => mixed,
  placeholder?: string,
  renderItem: ItemRenderer,
  renderMenu?: MenuRenderer,
  renderTag: TagRenderer,
  rootRef?: (ref: ?HTMLElement) => mixed,
  value: Array<Item>,
}

type State = {
  filteredItems: Array<Item>,
  focusedItemIndex: ?number,
  inputValue: string,
  isFocused: boolean,
  isOpen: boolean,
}

class MultiSelect extends React.Component<Props, State> {
  static defaultProps = {
    disabled: false,
    filterItems: (items: Array<Item>) => items,
    minimumInput: 0,
  }

  constructor(props: Props, ...args: Array<Object>) {
    super(props, ...args)

    this.state = {
      filteredItems: props.items,
      focusedItemIndex: null,
      inputValue: '',
      isFocused: false,
      isOpen: false,
    }

    const id = generateId('rechoice')

    this.id = id
    this.controlsId = `${id}-listbox`
    this.classes = getClasses(props.classes)
  }

  blurTimer: ?TimeoutID
  focusedItemNode: ?HTMLElement
  inputMirrorNode: ?HTMLElement
  inputNode: ?HTMLInputElement
  menuNode: ?HTMLElement
  rootNode: ?HTMLElement

  id: string
  controlsId: string
  classes: ClassMap

  componentDidMount() {
    addClickListener(this.onDocumentClick)
  }

  componentWillReceiveProps(nextProps: Props) {
    if (
      nextProps.items !== this.props.items ||
      nextProps.value !== this.props.value
    ) {
      this.setState(state => ({
        filteredItems: nextProps.filterItems(
          nextProps.items,
          state.inputValue,
          nextProps.value
        ),
      }))
    }
  }

  componentDidUpdate() {
    const { inputNode, inputMirrorNode } = this

    if (!inputNode || !inputMirrorNode) {
      return
    }

    inputNode.style.width = ''
    inputNode.style.width = `${inputMirrorNode.clientWidth}px`
  }

  componentWillUnmount() {
    removeClickListener(this.onDocumentClick)
    this.clearBlurTimer()
  }

  setInputRef = (ref: ?HTMLInputElement) => {
    if (this.props.inputRef) {
      this.props.inputRef(ref)
    }

    this.inputNode = ref
  }

  setFocusedItemRef = (ref: ?HTMLElement) => {
    this.focusedItemNode = ref
  }

  setInputMirrorRef = (ref: ?HTMLElement) => {
    this.inputMirrorNode = ref
  }

  setRootNode = (ref: ?HTMLElement) => {
    if (this.props.rootRef) {
      this.props.rootRef(ref)
    }

    this.rootNode = ref
  }

  setMenuRef = (ref: ?HTMLElement) => {
    this.menuNode = ref
  }

  close = () => {
    this.setState({
      isOpen: false,
      isFocused: false,
      focusedItemIndex: null,
    })
  }

  focusInput = () => {
    const { inputNode } = this
    if (inputNode) {
      inputNode.blur()
      inputNode.focus()
    }
  }

  clearBlurTimer = () => {
    if (this.blurTimer) {
      clearTimeout(this.blurTimer)
      this.blurTimer = null
    }
  }

  onDocumentClick = (e: Event) => {
    const { target } = e
    const { rootNode, menuNode } = this

    if (target instanceof Node && rootNode && menuNode) {
      if (!contains(rootNode, target) && !contains(menuNode, target)) {
        // close menu when clicking outside
        this.close()
      } else {
        // stop menu click from blurring the input
        e.preventDefault()
      }
    }
  }

  onInputBlur = () => {
    this.clearBlurTimer()
    if (this.props.onBlur) {
      this.props.onBlur()
    }
    this.blurTimer = setTimeout(this.close, 10)
  }

  onInputFocus = () => {
    this.clearBlurTimer()
    if (this.props.onFocus) {
      this.props.onFocus()
    }
    this.setState(state => focusInput(state, this.props), this.scrollToSelected)
  }

  onInputChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const inputValue = e.target.value

    if (this.props.onInputChange) {
      this.props.onInputChange(inputValue)
    }

    this.setState(state => updateInputValue(state, this.props, inputValue))
  }

  onKeyDown = (e: SyntheticKeyboardEvent<>) => {
    const {
      filteredItems,
      focusedItemIndex,
      inputValue,
      isFocused,
    } = this.state
    const { filterItems, loadingState, onChange, items, value } = this.props

    if (e.keyCode === keyCodes.BACKSPACE && !inputValue) {
      const newValue = value.slice(0, -1)
      onChange(newValue)
      this.setState(state => ({
        filteredItems: filterItems(items, state.inputValue, newValue),
      }))
    } else if (e.keyCode === keyCodes.ENTER) {
      if (loadingState === 'loading') {
        e.preventDefault()
        return
      }

      if (
        typeof focusedItemIndex === 'number' &&
        isFocused &&
        this.getIsOpen()
      ) {
        this.toggleItem(filteredItems[focusedItemIndex])
      } else {
        this.setState(state => ({
          isOpen: true,
          filteredItems: filterItems(items, state.inputValue, value),
        }))
      }
    } else if (e.keyCode === keyCodes.UP) {
      this.focusPreviousItem()
    } else if (e.keyCode === keyCodes.DOWN) {
      this.focusNextItem()
    } else {
      return
    }

    e.preventDefault()
  }

  onHoverItem = (i: number) => {
    this.setState({
      focusedItemIndex: i,
    })
  }

  toggleItem = (item: ?Item) => {
    if (!item) {
      return
    }

    const { onChange, itemKey, value } = this.props
    const itemId = item[itemKey]
    const itemIndex = value.findIndex(o => o[itemKey] === itemId)

    const newValue =
      itemIndex === -1
        ? [...value, item]
        : value.filter((o, i) => i !== itemIndex)

    onChange(newValue)

    this.setState(
      {
        inputValue: '',
        isOpen: false,
      },
      this.focusInput
    )
  }

  focusPreviousItem() {
    this.setState(
      state => focusPreviousItem(state, this.props),
      this.scrollToSelected
    )
  }

  focusNextItem() {
    this.setState(
      state => focusNextItem(state, this.props),
      this.scrollToSelected
    )
  }

  scrollToSelected = () => {
    const { focusedItemNode, menuNode } = this

    if (!menuNode || !focusedItemNode) {
      return
    }

    scrollIntoView(focusedItemNode, menuNode)
  }

  isItemSelected = (item: Item) => {
    const { value, itemKey } = this.props
    return Boolean(
      value.find(selectedItem => selectedItem[itemKey] === item[itemKey])
    )
  }

  getItemId = (index: number) => {
    return `${this.id}-item-${index}`
  }

  getIsOpen = () => {
    return typeof this.props.isOpen === 'undefined'
      ? this.state.isOpen
      : this.props.isOpen
  }

  render() {
    const {
      disabled,
      loadingIndicator,
      itemKey,
      loadingState,
      placeholder,
      renderItem,
      renderMenu,
      renderTag,
      value,
    } = this.props

    const {
      filteredItems,
      focusedItemIndex,
      inputValue,
      isFocused,
    } = this.state

    const selectClasses = cx(this.classes.root, {
      [this.classes.focused]: isFocused,
      [this.classes.disabled]: disabled,
    })

    const isOpen = this.getIsOpen()

    const inputMirrorText = value.length
      ? inputValue
      : inputValue || placeholder

    const focusedItemId =
      typeof focusedItemIndex === 'number'
        ? this.getItemId(focusedItemIndex)
        : null

    return (
      <div className={selectClasses}>
        <div
          className={this.classes.inputWrapper}
          onClick={this.focusInput}
          ref={this.setRootNode}
          role="presentation"
        >
          <Tags
            className={this.classes.tags}
            itemKey={itemKey}
            removeTag={this.toggleItem}
            renderTag={renderTag}
            value={value}
          />
          <input
            aria-activedescendant={focusedItemId}
            aria-autocomplete="list"
            aria-owns={this.controlsId}
            autoComplete="off"
            className={this.classes.input}
            disabled={disabled}
            onBlur={this.onInputBlur}
            onChange={this.onInputChange}
            onFocus={this.onInputFocus}
            onKeyDown={this.onKeyDown}
            placeholder={value.length ? null : placeholder}
            ref={this.setInputRef}
            type="text"
            value={inputValue}
          />
          <span
            className={this.classes.inputMirror}
            ref={this.setInputMirrorRef}
          >
            {inputMirrorText}{' '}
          </span>
        </div>
        <Menu
          classes={this.classes}
          controlsId={this.controlsId}
          focusedItemIndex={focusedItemIndex}
          focusedItemRef={this.setFocusedItemRef}
          getItemId={this.getItemId}
          inputValue={inputValue}
          isItemSelected={this.isItemSelected}
          isOpen={isOpen}
          itemKey={itemKey}
          items={filteredItems}
          loadingIndicator={loadingIndicator}
          menuRef={this.setMenuRef}
          onHoverItem={this.onHoverItem}
          onSelect={this.toggleItem}
          renderItem={renderItem}
          renderMenu={renderMenu}
          state={loadingState}
        />
      </div>
    )
  }
}

export default MultiSelect
