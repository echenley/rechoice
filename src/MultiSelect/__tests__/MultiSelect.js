/* eslint-disable max-nested-callbacks */

jest.mock('../../util/clickListener')
jest.mock('../../util/contains')
jest.mock('../../util/scrollIntoView')

jest.useFakeTimers()

import * as React from 'react'
import { shallow } from 'enzyme'
import { addClickListener, removeClickListener } from '../../util/clickListener'
import contains from '../../util/contains'
import keyCodes from '../../util/keyCodes'
import scrollIntoView from '../../util/scrollIntoView'
import MultiSelect from '../MultiSelect'
import Menu from '../../Menu'

contains.mockImplementation(() => false)

const item3 = { id: 3 }

// prettier-ignore
const items = [
  { id: 1 },
  { id: 2 },
  item3,
  { id: 4 },
  { id: 5 },
]

const renderComponent = props => {
  return shallow(
    <MultiSelect
      itemKey="id"
      items={items}
      loadingIndicator={<div className="loading-indicator" />}
      onBlur={() => {}}
      onChange={() => {}}
      onFocus={() => {}}
      placeholder="Placeholderâ€¦"
      renderItem={() => {}}
      renderSelected={() => {}}
      value={[item3]}
      {...props}
    />
  )
}

describe('MultiSelect', () => {
  afterEach(() => {
    clearTimeout.mockClear()
  })

  it('renders correctly', () => {
    expect(renderComponent().getElement()).toMatchSnapshot()
  })

  it('renders correctly when isOpen state to be overridden by props.isOpen', () => {
    const component = renderComponent({ isOpen: true })
    expect(component.find(Menu).prop('isOpen')).toBe(true)
  })

  it('adds "disabled" class and disables input when disabled', () => {
    const component = renderComponent({ disabled: true })
    expect(component.hasClass('disabled')).toBe(true)
    expect(component.find('input').prop('disabled')).toBe(true)
  })

  it('adds focused class when isFocused and calls onFocus props if provided', () => {
    const onFocus = jest.fn()
    const component = renderComponent({ onFocus })
    component.instance().onInputFocus()
    component.update()
    expect(component.hasClass('focused')).toBe(true)
    expect(onFocus).toHaveBeenCalled()
  })

  describe('Methods', () => {
    describe('componentDidMount', () => {
      it('calls addClickListener', () => {
        const component = renderComponent()
        component.instance().componentDidMount()
        expect(addClickListener).toHaveBeenCalledWith(
          component.instance().onDocumentClick
        )
      })
    })

    describe('componentWillReceiveProps', () => {
      it('updates filteredItems if props.items changes', () => {
        const component = renderComponent()
        const newItems = [...items, { id: 6 }]
        component.setProps({
          items: newItems,
        })

        expect(component.state('filteredItems')).toEqual(newItems)
      })
    })

    describe('componentDidUpdate', () => {
      it('does not error if inputNode or inputMirrorNode are not defined', () => {
        const component = renderComponent()
        component.setState({ inputValue: 'test' })
      })

      it('sets inputNode width to inputMirrorNode width', () => {
        const component = renderComponent()
        const instance = component.instance()

        instance.inputNode = { style: { width: '0px' } }
        instance.inputMirrorNode = { clientWidth: 200 }

        component.setState({
          inputValue: 'test',
        })

        expect(instance.inputNode.style.width).toBe('200px')
      })
    })

    describe('componentWillUnmount', () => {
      it('calls removeClickListener', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.clearBlurTimer = jest.fn()
        instance.componentWillUnmount()

        expect(removeClickListener).toHaveBeenCalledWith(
          instance.onDocumentClick
        )
        expect(instance.clearBlurTimer).toHaveBeenCalled()
      })
    })

    describe('close', () => {
      it('sets closed state', () => {
        const component = renderComponent()
        component.setState({
          isOpen: true,
          isFocused: true,
        })

        component.instance().close()

        expect(component.state('isOpen')).toBe(false)
        expect(component.state('isFocused')).toBe(false)
      })
    })

    describe('focusInput', () => {
      it('calls .focus() on input node', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.inputNode = { focus: jest.fn(), blur: jest.fn() }
        instance.focusInput()
        expect(instance.inputNode.focus).toHaveBeenCalled()
        expect(instance.inputNode.blur).toHaveBeenCalled()
      })

      it('does not call .focus() if input node is null', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.inputNode = undefined
        expect(() => instance.focusInput()).not.toThrow()
      })
    })

    describe('clearBlurTimer', () => {
      it('calls clearTimeout on this.blurTimer', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.blurTimer = 111
        instance.clearBlurTimer()
        expect(clearTimeout).toHaveBeenCalledWith(111)
      })

      it('is a no-op if this.blurTimer is null', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.blurTimer = null
        instance.clearBlurTimer()
        expect(clearTimeout).not.toHaveBeenCalled()
      })
    })

    describe('onDocumentClick', () => {
      it('calls this.close if click target was not within the select or popover', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.close = jest.fn()
        instance.menuNode = {}
        instance.rootNode = {}
        instance.onDocumentClick({ target: document.createElement('div') })
        expect(instance.close).toHaveBeenCalled()
      })

      it('does not call this.close if click target was outside the select and popover', () => {
        contains.mockImplementationOnce(() => true)
        const component = renderComponent()
        const instance = component.instance()
        instance.close = jest.fn()
        instance.onDocumentClick({ target: {} })
        expect(instance.close).not.toHaveBeenCalled()
      })
    })

    describe('onInputBlur', () => {
      it('calls onBlur prop if provided', () => {
        const onBlur = jest.fn()
        const component = renderComponent({
          onBlur,
        })
        const instance = component.instance()
        instance.close = jest.fn()
        instance.clearBlurTimer = jest.fn()
        instance.onInputBlur()

        expect(onBlur).toHaveBeenCalled()
      })

      it('calls clearBlurTimer and close after a setTimeout', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.close = jest.fn()
        instance.clearBlurTimer = jest.fn()
        instance.onInputBlur()

        expect(instance.clearBlurTimer).toHaveBeenCalled()
        expect(instance.close).not.toHaveBeenCalled()

        jest.runAllTimers()

        expect(instance.close).toHaveBeenCalled()
        expect(Boolean(instance.blurTimer)).toBe(true)
      })
    })

    describe('onInputFocus', () => {
      it('calls clearBlurTimer and updates state', () => {
        const component = renderComponent()
        const instance = component.instance()
        instance.clearBlurTimer = jest.fn()
        component.find('input').prop('onFocus')()
        expect(instance.clearBlurTimer).toHaveBeenCalled()
        expect(component.state()).toMatchSnapshot()
      })
    })

    describe('onKeyDown', () => {
      it('BACKSPACE removes the last tag if input is empty', () => {
        const onChange = jest.fn()
        const component = renderComponent({
          filterItems: () => [],
          onChange,
          value: [{ id: 1 }, { id: 2 }],
        })

        const instance = component.instance()

        component.setState({
          filteredItems: [{ id: 1 }, { id: 2 }, { id: 3 }],
        })

        const e = {
          keyCode: keyCodes.BACKSPACE,
          preventDefault: jest.fn(),
        }

        instance.onKeyDown(e)

        expect(onChange).toHaveBeenCalledWith([{ id: 1 }])
        expect(component.state('filteredItems')).toEqual([])
      })

      it('ENTER opens the popover and updates filteredItems if popover is closed', () => {
        const component = renderComponent({
          filterItems: () => [],
        })
        const instance = component.instance()

        component.setState({ isOpen: false })

        const e = {
          keyCode: keyCodes.ENTER,
          preventDefault: jest.fn(),
        }

        instance.onKeyDown(e)

        expect(component.state('isOpen')).toBe(true)
        expect(component.state('filteredItems')).toEqual([])
        expect(e.preventDefault).toHaveBeenCalled()
      })

      it('ENTER calls this.toggleItem if popover is open and focused', () => {
        const component = renderComponent()
        const instance = component.instance()

        component.setState({
          focusedItemIndex: 2,
          inputValue: 'test',
          isOpen: true,
          isFocused: true,
        })

        const e = {
          keyCode: keyCodes.ENTER,
          preventDefault: jest.fn(),
        }

        instance.toggleItem = jest.fn()
        instance.onKeyDown(e)

        expect(instance.toggleItem).toHaveBeenCalledWith(item3)
        expect(e.preventDefault).toHaveBeenCalled()
      })

      it('ENTER does nothing when loading', () => {
        const component = renderComponent({
          loadingState: 'loading',
        })
        const instance = component.instance()

        component.setState({
          focusedItemIndex: 2,
          inputValue: 'test',
          isOpen: true,
          isFocused: true,
        })

        const e = {
          keyCode: keyCodes.ENTER,
          preventDefault: jest.fn(),
        }

        instance.onSelect = jest.fn()
        instance.onKeyDown(e)

        expect(instance.onSelect).not.toHaveBeenCalled()
        expect(e.preventDefault).toHaveBeenCalled()
      })

      it('UP focuses previous item and calls scrollIntoView', () => {
        const component = renderComponent()
        const instance = component.instance()

        const e = {
          keyCode: keyCodes.UP,
          preventDefault: jest.fn(),
        }

        instance.menuNode = { id: 'menuNode' }
        instance.focusedItemNode = { id: 'focusedItemNode' }
        instance.focusPreviousItem = jest.fn(instance.focusPreviousItem)
        instance.onKeyDown(e)

        expect(instance.focusPreviousItem).toHaveBeenCalledWith()
        expect(e.preventDefault).toHaveBeenCalled()
        expect(scrollIntoView).toHaveBeenCalledWith(
          instance.focusedItemNode,
          instance.menuNode
        )
      })

      it('DOWN focuses next item and calls scrollIntoView', () => {
        const component = renderComponent()
        const instance = component.instance()

        const e = {
          keyCode: keyCodes.DOWN,
          preventDefault: jest.fn(),
        }

        instance.menuNode = { id: 'menuNode' }
        instance.focusedItemNode = { id: 'focusedItemNode' }
        instance.focusNextItem = jest.fn(instance.focusNextItem)
        instance.onKeyDown(e)

        expect(instance.focusNextItem).toHaveBeenCalledWith()
        expect(e.preventDefault).toHaveBeenCalled()
        expect(scrollIntoView).toHaveBeenCalledWith(
          instance.focusedItemNode,
          instance.menuNode
        )
      })

      it('does not preventDefault for any other keys', () => {
        const component = renderComponent()
        const instance = component.instance()

        const e = {
          keyCode: 100,
          preventDefault: jest.fn(),
        }

        instance.onKeyDown(e)
        expect(e.preventDefault).not.toHaveBeenCalled()
      })
    })

    describe('toggleItem', () => {
      it('does nothing if item is null', () => {
        const onChange = jest.fn()
        const component = renderComponent({ onChange })
        const instance = component.instance()
        instance.toggleItem(null)
        expect(onChange).not.toHaveBeenCalled()
      })

      it('calls onChange with new value, clears input, closes popover, and resets filteredItems', () => {
        const onChange = jest.fn()
        const items = [{ id: 1 }, { id: 2 }, item3, { id: 4 }]
        const component = renderComponent({
          onChange,
          items,
          value: [],
        })
        const instance = component.instance()
        instance.focusInput = jest.fn()

        component.setState({
          filteredItems: [item3],
          inputValue: 'test',
          isOpen: true,
        })

        instance.toggleItem(item3)

        expect(onChange).toHaveBeenCalledWith([item3])
        expect(instance.focusInput).toHaveBeenCalled()
        expect(component.state('inputValue')).toBe('')
        expect(component.state('isOpen')).toBe(false)
      })

      it('removes old item from value if already selected', () => {
        const onChange = jest.fn()
        const items = [{ id: 1 }, { id: 2 }, item3, { id: 4 }]
        const component = renderComponent({
          onChange,
          items,
          value: [item3],
        })
        const instance = component.instance()
        instance.focusInput = jest.fn()

        component.setState({
          filteredItems: [item3],
          inputValue: 'test',
          isOpen: true,
        })

        instance.toggleItem(item3)

        expect(onChange).toHaveBeenCalledWith([])
        expect(instance.focusInput).toHaveBeenCalled()
        expect(component.state('inputValue')).toBe('')
        expect(component.state('isOpen')).toBe(false)
      })
    })

    describe('onInputChange', () => {
      it('updates state with new input value', () => {
        const onInputChange = jest.fn()
        const component = renderComponent({
          onInputChange,
        })
        const instance = component.instance()
        instance.onInputChange({
          target: {
            value: 'test',
          },
        })

        expect(onInputChange).toHaveBeenCalledWith('test')
        expect(component.state()).toMatchSnapshot()
      })
    })

    describe('isItemSelected', () => {
      it('returns true if item matches props.value', () => {
        const component = renderComponent()
        expect(component.instance().isItemSelected({ id: 3 })).toBe(true)
      })

      it('returns false if item does not match props.value', () => {
        const component = renderComponent()
        expect(component.instance().isItemSelected({ id: 2 })).toBe(false)
      })
    })
  })
})
