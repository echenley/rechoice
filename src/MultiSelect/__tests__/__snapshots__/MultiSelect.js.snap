// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect Methods onInputChange updates state with new input value 1`] = `
Object {
  "filteredItems": Array [
    Object {
      "id": 1,
    },
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
    Object {
      "id": 5,
    },
  ],
  "focusedItemIndex": 0,
  "inputValue": "test",
  "isFocused": false,
  "isOpen": true,
}
`;

exports[`MultiSelect Methods onInputFocus calls clearBlurTimer and updates state 1`] = `
Object {
  "filteredItems": Array [
    Object {
      "id": 1,
    },
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
    Object {
      "id": 5,
    },
  ],
  "focusedItemIndex": 0,
  "inputValue": "",
  "isFocused": true,
  "isOpen": true,
}
`;

exports[`MultiSelect renders correctly 1`] = `
<div
  className="rechoice"
>
  <div
    className="rechoice-input-wrapper"
    onClick={[Function]}
    role="presentation"
  >
    <Tags
      className="rechoice-tags"
      itemKey="id"
      removeTag={[Function]}
      renderTag={undefined}
      value={
        Array [
          Object {
            "id": 3,
          },
        ]
      }
    />
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-owns="rechoice-0-listbox"
      autoComplete="off"
      className="rechoice-input"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder={null}
      type="text"
      value=""
    />
    <span
      className="rechoice-input-mirror"
    >
      
       
    </span>
  </div>
  <Menu
    classes={
      Object {
        "disabled": "disabled",
        "focused": "focused",
        "input": "rechoice-input",
        "inputMirror": "rechoice-input-mirror",
        "inputWrapper": "rechoice-input-wrapper",
        "item": "rechoice-item",
        "menu": "rechoice-menu",
        "root": "rechoice",
        "selected": "selected",
        "tags": "rechoice-tags",
        "value": "rechoice-value",
      }
    }
    controlsId="rechoice-0-listbox"
    focusedItemIndex={null}
    focusedItemRef={[Function]}
    getItemId={[Function]}
    inputValue=""
    isItemSelected={[Function]}
    isOpen={false}
    itemKey="id"
    items={
      Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
        Object {
          "id": 4,
        },
        Object {
          "id": 5,
        },
      ]
    }
    loadingIndicator={
      <div
        className="loading-indicator"
      />
    }
    menuRef={[Function]}
    onHoverItem={[Function]}
    onSelect={[Function]}
    renderItem={[Function]}
    renderMenu={undefined}
    state={undefined}
  />
</div>
`;
